File: App/DivyaApp/ios/DivyaApp/NativeWSModule.swift
import Foundation
import React

@objc(NativeWS)
class NativeWSModule: RCTEventEmitter {
  private var session: URLSession?
  private var task: URLSessionWebSocketTask?
  private var nativeAudioPlaybackEnabled = false
  private var needsPcmReinit = false
  private var batch: [[String: Any]] = []
  private var batchTimer: DispatchSourceTimer?
  private let batchIntervalMs: UInt64 = 50

  override init() {
    super.init()
    NativeWSModuleHolder.instance = self
  }

  @objc
  override static func requiresMainQueueSetup() -> Bool {
    return false
  }

  override func supportedEvents() -> [String]! {
    return [
      "NativeWSOpen",
      "NativeWSClose",
      "NativeWSError",
      "NativeWSBatchParsed",
      "NativeWSParsedImmediate",
      "NativeWSBinary"
    ]
  }

  @objc(connect:headers:)
  func connect(_ url: String, headers: NSDictionary?) {
    disconnect()
    guard let requestUrl = URL(string: url) else { return }
    var request = URLRequest(url: requestUrl)
    if let headers = headers {
      for (key, value) in headers {
        if let k = key as? String, let v = value as? String {
          request.setValue(v, forHTTPHeaderField: k)
        }
      }
    }
    let configuration = URLSessionConfiguration.default
    session = URLSession(configuration: configuration)
    task = session?.webSocketTask(with: request)
    task?.resume()
    sendEvent(withName: "NativeWSOpen", body: nil)
    receiveLoop()
  }

  @objc
  func disconnect() {
    task?.cancel(with: .normalClosure, reason: nil)
    task = nil
    session?.invalidateAndCancel()
    session = nil
  }

  @objc
  func sendText(_ text: String) {
    task?.send(.string(text)) { _ in }
  }

  @objc
  func sendBinaryBase64(_ base64: String) {
    if let data = Data(base64Encoded: base64) {
      task?.send(.data(data)) { _ in }
    }
  }

  @objc
  func setNativeAudioPlaybackEnabled(_ enabled: Bool) {
    nativeAudioPlaybackEnabled = enabled
  }

  func sendAudioBase64FromNative(_ base64: String) {
    let payload: [String: Any] = ["type": "audio", "audio_data": base64]
    if let data = try? JSONSerialization.data(withJSONObject: payload),
       let text = String(data: data, encoding: .utf8) {
      task?.send(.string(text)) { _ in }
    }
  }

  private func receiveLoop() {
    task?.receive { [weak self] result in
      guard let self = self else { return }
      switch result {
      case .failure(let error):
        self.sendEvent(withName: "NativeWSError", body: ["message": error.localizedDescription])
      case .success(let message):
        switch message {
        case .string(let text):
          self.handleMessage(text)
        case .data(let data):
          let base64 = data.base64EncodedString()
          self.sendEvent(withName: "NativeWSBinary", body: base64)
        @unknown default:
          break
        }
      }
      self.receiveLoop()
    }
  }

  private func handleMessage(_ text: String) {
    guard let jsonData = text.data(using: .utf8),
          let obj = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any]
    else {
      enqueue(["type": "unknown", "raw": text])
      return
    }
    let type = (obj["type"] as? String) ?? "unknown"

    if type == "interrupted" && nativeAudioPlaybackEnabled {
      PcmPlayerModuleHolder.instance?.stop(nil)
      needsPcmReinit = true
      NativeWSModuleHolder.botPlaying = false
      NativeWSModuleHolder.dropAudioUntilTurnComplete = true
    }
    if type == "turn_complete" {
      NativeWSModuleHolder.botPlaying = false
      NativeWSModuleHolder.dropAudioUntilTurnComplete = false
    }
    if type == "input_transcription" && NativeWSModuleHolder.botPlaying {
      PcmPlayerModuleHolder.instance?.stop(nil)
      needsPcmReinit = true
      NativeWSModuleHolder.botPlaying = false
      NativeWSModuleHolder.dropAudioUntilTurnComplete = true
    }

    if type == "audio_chunk" && nativeAudioPlaybackEnabled {
      NativeWSModuleHolder.botPlaying = true
      if NativeWSModuleHolder.dropAudioUntilTurnComplete {
        enqueue(parseMessage(obj, includeAudioData: false))
        return
      }
      let audioData = obj["audio_data"] as? String ?? ""
      let mimeType = obj["mime_type"] as? String ?? ""
      let rate = parseRate(mimeType)
      if !audioData.isEmpty {
        if needsPcmReinit || rate != nil {
          PcmPlayerModuleHolder.instance?.initPlayer(rate ?? 24000, promise: nil)
          needsPcmReinit = false
        }
        PcmPlayerModuleHolder.instance?.writeBase64(audioData, promise: nil)
      }
      enqueue(parseMessage(obj, includeAudioData: false))
      return
    }

    let parsed = parseMessage(obj, includeAudioData: true)
    if type == "output_transcription" {
      sendEvent(withName: "NativeWSParsedImmediate", body: parsed)
      return
    }
    enqueue(parsed)
  }

  private func parseRate(_ mimeType: String) -> Int? {
    let pattern = "rate=([0-9]+)"
    guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
    let range = NSRange(location: 0, length: mimeType.count)
    guard let match = regex.firstMatch(in: mimeType, range: range),
          match.numberOfRanges > 1,
          let rateRange = Range(match.range(at: 1), in: mimeType)
    else { return nil }
    return Int(mimeType[rateRange])
  }

  private func parseMessage(_ obj: [String: Any], includeAudioData: Bool) -> [String: Any] {
    var map: [String: Any] = ["type": obj["type"] as? String ?? "unknown"]
    if let text = obj["text"] as? String { map["text"] = text }
    if let tool = obj["tool"] as? String { map["tool"] = tool }
    if let phase = obj["phase"] as? String { map["phase"] = phase }
    if let label = obj["label"] as? String { map["label"] = label }
    if let summary = obj["summary"] as? String { map["summary"] = summary }
    if let duration = obj["duration_ms"] as? Int { map["duration_ms"] = duration }
    if includeAudioData, let audio = obj["audio_data"] as? String { map["audio_data"] = audio }
    if let chunk = obj["chunk_number"] as? Int { map["chunk_number"] = chunk }
    if let size = obj["chunk_size"] as? Int { map["chunk_size"] = size }
    if let mime = obj["mime_type"] as? String { map["mime_type"] = mime }
    if let err = obj["error"] as? String { map["error"] = err }
    if let data = obj["data"] {
      map["data"] = data
    }
    return map
  }

  private func enqueue(_ message: [String: Any]) {
    batch.append(message)
    startBatching()
  }

  private func startBatching() {
    if batchTimer != nil { return }
    let timer = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
    timer.schedule(deadline: .now() + .milliseconds(Int(batchIntervalMs)), repeating: .milliseconds(Int(batchIntervalMs)))
    timer.setEventHandler { [weak self] in
      guard let self = self else { return }
      if self.batch.isEmpty {
        self.batchTimer?.cancel()
        self.batchTimer = nil
        return
      }
      let out = self.batch
      self.batch.removeAll()
      self.sendEvent(withName: "NativeWSBatchParsed", body: out)
    }
    batchTimer = timer
    timer.resume()
  }
}

File: App/DivyaApp/android/app/src/main/java/com/divyaapp/ws/NativeWSModule.kt
package com.divyaapp.ws

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.Arguments
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import okhttp3.Headers
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okhttp3.logging.HttpLoggingInterceptor
import okio.ByteString
import okio.ByteString.Companion.decodeBase64
import org.json.JSONArray
import org.json.JSONObject
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.divyaapp.pcm.PcmPlayerModuleHolder
import com.divyaapp.mic.NativeMicModuleHolder

class NativeWSModule(private val reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  private var client: OkHttpClient? = null
  private var socket: WebSocket? = null
  private val handler = Handler(Looper.getMainLooper())
  private val batch: MutableList<WritableMap> = ArrayList()
  private var batching = false
  private val batchIntervalMs = 50L
  private var nativeAudioPlaybackEnabled = false
  private val logTag = "NativeWS"
  private var needsPcmReinit = false

  init {
    NativeWSModuleHolder.INSTANCE = this
  }

  override fun getName(): String = "NativeWS"

  @ReactMethod
  fun addListener(eventName: String) {
    // Required for NativeEventEmitter; no-op.
  }

  @ReactMethod
  fun removeListeners(count: Int) {
    // Required for NativeEventEmitter; no-op.
  }

  @ReactMethod
  fun connect(url: String, headers: ReadableMap?) {
    disconnect()
    val headerBuilder = Headers.Builder()
    headers?.keySetIterator()?.let { iterator ->
      while (iterator.hasNextKey()) {
        val key = iterator.nextKey()
        val value = headers.getString(key)
        if (!value.isNullOrEmpty()) {
          headerBuilder.add(key, value)
        }
      }
    }
    val request = Request.Builder()
      .url(url)
      .headers(headerBuilder.build())
      .build()
    val logging = HttpLoggingInterceptor { message ->
      Log.d(logTag, "http $message")
    }.apply {
      level = HttpLoggingInterceptor.Level.HEADERS
    }
    client = OkHttpClient.Builder()
      .addInterceptor(logging)
      .build()
    socket = client!!.newWebSocket(request, NativeWSListener())
  }

  @ReactMethod
  fun sendText(text: String) {
    socket?.send(text)
  }

  fun sendAudioBase64FromNative(base64: String) {
    val payload = JSONObject()
    payload.put("type", "audio")
    payload.put("audio_data", base64)
    socket?.send(payload.toString())
  }

  fun onNativeBargeIn(reason: String) {
    if (NativeWSModuleHolder.interruptedEmitted) return
    needsPcmReinit = true
    val map = Arguments.createMap()
    map.putString("type", "interrupted")
    map.putString("reason", reason)
    map.putDouble("timestamp", System.currentTimeMillis().toDouble())
    enqueueMessage(map)
    NativeWSModuleHolder.interruptedEmitted = true
  }

  @ReactMethod
  fun sendBinaryBase64(base64: String) {
    socket?.send(base64.decodeBase64() ?: return)
  }

  @ReactMethod
  fun setNativeAudioPlaybackEnabled(enabled: Boolean) {
    nativeAudioPlaybackEnabled = enabled
    Log.d(logTag, "native audio playback enabled=$enabled")
  }

  @ReactMethod
  fun disconnect() {
    socket?.close(1000, "client_disconnect")
    socket = null
    client?.dispatcher?.executorService?.shutdown()
    client = null
  }

  private fun emit(event: String, payload: Any?) {
    reactContext
      .getJSModule(RCTDeviceEventEmitter::class.java)
      .emit(event, payload)
  }

  private fun enqueueMessage(msg: WritableMap) {
    synchronized(batch) {
      batch.add(msg)
    }
    startBatching()
  }

  private fun startBatching() {
    if (batching) return
    batching = true
    handler.post(batchRunnable)
  }

  private val batchRunnable = object : Runnable {
    override fun run() {
      val out: WritableArray = Arguments.createArray()
      synchronized(batch) {
        batch.forEach { out.pushMap(it) }
        batch.clear()
      }
      if (out.size() > 0) {
        emit("NativeWSBatchParsed", out)
        handler.postDelayed(this, batchIntervalMs)
      } else {
        batching = false
      }
    }
  }

  private fun parseMessage(text: String): WritableMap {
    val map = Arguments.createMap()
    try {
      val obj = JSONObject(text)
      return parseMessage(obj, true)
    } catch (_: Exception) {
      map.putString("type", "unknown")
      map.putString("raw", text)
    }
    return map
  }

  private fun parseMessage(obj: JSONObject, includeAudioData: Boolean): WritableMap {
    val map = Arguments.createMap()
    val type = obj.optString("type", "unknown")
    map.putString("type", type)
    if (obj.has("text")) map.putString("text", obj.optString("text", ""))
    if (obj.has("tool")) map.putString("tool", obj.optString("tool", ""))
    if (obj.has("phase")) map.putString("phase", obj.optString("phase", ""))
    if (obj.has("label")) map.putString("label", obj.optString("label", ""))
    if (obj.has("summary")) map.putString("summary", obj.optString("summary", ""))
    if (obj.has("duration_ms")) map.putInt("duration_ms", obj.optInt("duration_ms"))
    if (includeAudioData && obj.has("audio_data")) {
      map.putString("audio_data", obj.optString("audio_data", ""))
    }
    if (obj.has("chunk_number")) map.putInt("chunk_number", obj.optInt("chunk_number"))
    if (obj.has("chunk_size")) map.putInt("chunk_size", obj.optInt("chunk_size"))
    if (obj.has("mime_type")) map.putString("mime_type", obj.optString("mime_type", ""))
    if (obj.has("error")) map.putString("error", obj.optString("error", ""))
    if (obj.has("data")) {
      val data = obj.opt("data")
      when (data) {
        is JSONObject -> map.putMap("data", jsonToMap(data))
        is JSONArray -> map.putArray("data", jsonToArray(data))
        else -> map.putString("data", data?.toString() ?: "")
      }
    }
    return map
  }

  private fun jsonToMap(obj: JSONObject): WritableMap {
    val map = Arguments.createMap()
    val keys = obj.keys()
    while (keys.hasNext()) {
      val key = keys.next()
      val value = obj.opt(key)
      when (value) {
        is JSONObject -> map.putMap(key, jsonToMap(value))
        is JSONArray -> map.putArray(key, jsonToArray(value))
        is Boolean -> map.putBoolean(key, value)
        is Int -> map.putInt(key, value)
        is Long -> map.putDouble(key, value.toDouble())
        is Double -> map.putDouble(key, value)
        else -> map.putString(key, value?.toString() ?: "")
      }
    }
    return map
  }

  private fun jsonToArray(arr: JSONArray): WritableArray {
    val array = Arguments.createArray()
    for (i in 0 until arr.length()) {
      val value = arr.opt(i)
      when (value) {
        is JSONObject -> array.pushMap(jsonToMap(value))
        is JSONArray -> array.pushArray(jsonToArray(value))
        is Boolean -> array.pushBoolean(value)
        is Int -> array.pushInt(value)
        is Long -> array.pushDouble(value.toDouble())
        is Double -> array.pushDouble(value)
        else -> array.pushString(value?.toString() ?: "")
      }
    }
    return array
  }

  private inner class NativeWSListener : WebSocketListener() {
    private fun logStreamEvent(type: String, obj: JSONObject) {
      when (type) {
        "audio_chunk" -> {
          val chunkNum = obj.optInt("chunk_number", -1)
          val chunkSize = obj.optInt("chunk_size", -1)
          val mimeType = obj.optString("mime_type", "")
          Log.d(logTag, "event audio_chunk chunk=$chunkNum size=$chunkSize mime=$mimeType")
        }
        "input_transcription", "output_transcription", "text_chunk" -> {
          val text = obj.optString("text", "")
          Log.d(logTag, "event $type text_len=${text.length}")
        }
        "tool_call_start", "tool_call_complete", "itinerary", "connected",
        "setup_complete", "turn_complete", "pong", "error", "interrupted" -> {
          Log.d(logTag, "event $type")
        }
        else -> {
          Log.d(logTag, "event $type")
        }
      }
    }

    override fun onOpen(webSocket: WebSocket, response: Response) {
      emit("NativeWSOpen", null)
    }

    override fun onMessage(webSocket: WebSocket, text: String) {
      val obj = JSONObject(text)
      val type = obj.optString("type", "unknown")
      logStreamEvent(type, obj)
      if (type == "interrupted" && nativeAudioPlaybackEnabled) {
        val pcmModule = PcmPlayerModuleHolder.getInstance()
        pcmModule?.stop(null)
        needsPcmReinit = true
        Log.w(logTag, "interrupted: stopping playback and marking reinit")
        NativeWSModuleHolder.botPlaying = false
        NativeWSModuleHolder.dropAudioUntilTurnComplete = true
      }
      if (type == "turn_complete") {
        NativeWSModuleHolder.botPlaying = false
        NativeWSModuleHolder.interruptedEmitted = false
        NativeWSModuleHolder.dropAudioUntilTurnComplete = false
      }
      if (type == "input_transcription" && NativeWSModuleHolder.botPlaying) {
        val pcmModule = PcmPlayerModuleHolder.getInstance()
        pcmModule?.stop(null)
        needsPcmReinit = true
        NativeWSModuleHolder.botPlaying = false
        NativeWSModuleHolder.dropAudioUntilTurnComplete = true
        Log.w(logTag, "input_transcription: stopping playback for new user turn")
      }
      if (type == "audio_chunk" && nativeAudioPlaybackEnabled) {
        NativeWSModuleHolder.botPlaying = true
        if (NativeWSModuleHolder.dropAudioUntilTurnComplete) {
          val parsed = parseMessage(obj, false)
          enqueueMessage(parsed)
          return
        }
        val audioData = obj.optString("audio_data", "")
        val mimeType = obj.optString("mime_type", "")
        val rateMatch = Regex("rate=(\\d+)").find(mimeType)
        val rate = rateMatch?.groups?.get(1)?.value?.toIntOrNull()
        if (audioData.isNotEmpty()) {
          val pcmModule = PcmPlayerModuleHolder.getInstance()
          if (needsPcmReinit || rate != null) {
            pcmModule?.init(rate ?: 24000, null)
            needsPcmReinit = false
          }
          if (pcmModule == null) {
            Log.w(logTag, "pcm module missing; cannot play audio")
          } else {
            Log.d(logTag, "writing audio_chunk bytes=${audioData.length}")
            pcmModule.writeBase64(audioData, null)
            Log.d(logTag, "played audio_chunk bytes=${audioData.length}")
          }
        }
        val parsed = parseMessage(obj, false)
        enqueueMessage(parsed)
        return
      }
      val parsed = parseMessage(obj, true)
      if (type == "output_transcription") {
        emit("NativeWSParsedImmediate", parsed)
        return
      }
      enqueueMessage(parsed)
    }

    override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
      emit("NativeWSBinary", bytes.base64())
    }

    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
      val payload = Arguments.createMap()
      payload.putInt("code", code)
      payload.putString("reason", reason)
      emit("NativeWSClose", payload)
    }

    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
      val payload = Arguments.createMap()
      payload.putInt("code", code)
      payload.putString("reason", reason)
      emit("NativeWSClose", payload)
    }

    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
      val payload = Arguments.createMap()
      payload.putString("message", t.message ?: "native_ws_error")
      payload.putString("cause", t.cause?.message ?: "")
      emit("NativeWSError", payload)
    }
  }
}
